screens/Search/index.tsx

```diff
@@ -18,75 +18,62 @@ import { User } from "@/types"
+
+  const { data: users = [] } = useQuery<User[]>({
+    queryKey: ["users"],
+    queryFn: async () => {
       const response = await api.get(`/users`)
-      const users = response.data.map((user: User) => ({
+      return response.data.map((user: User) => ({
         id: user.id,
         names: user.names,
         username: user.username,
         profilePictureUrl: user.profilePictureUrl,
         isFriend: user.isFriend,
       }))
-      setAllUsers(users)
-      setFilteredUsers(users)
-    } catch (error) {
-      console.error("error fetching users", error)
-    }
-  }, [])
-  useEffect(() => {
-    fetchUsers()
-  }, [fetchUsers])
+    },
+  })
+
+  const addFriendMutation = useMutation({
+    mutationFn: (friendId: string) => api.post("/friends", { friendId }),
+    onMutate: async (friendId) => {
+      await queryClient.cancelQueries({ queryKey: ["users"] })
+      const previousUsers = queryClient.getQueryData(["users"])
+      queryClient.setQueryData(["users"], (oldUsers: User[]) =>
+        oldUsers.map((user) =>
+          user.id === friendId ? { ...user, isFriend: true } : user,
+        ),
+      )
+      return { previousUsers }
+    },
+    onError: (err, friendId, context) => {
+      queryClient.setQueryData(["users"], context?.previousUsers)
+      console.warn("Error adding friend:", err)
+    },
+    onSettled: () => {
+      queryClient.invalidateQueries({ queryKey: ["users"] })
+      queryClient.invalidateQueries({ queryKey: ["friends"] })
+    },
+  })
+
+  const filteredUsers = users
+    .filter((user) => user.names.toLowerCase().includes(search.toLowerCase()))
+    .sort((a, b) => a.names.localeCompare(b.names))
 
   const handleSearch = (name: string) => {
     setSearch(name)
-    const filtered = allUsers
-      .filter((user) => user.names.toLowerCase().includes(name.toLowerCase()))
-      .sort((a, b) => a.names.localeCompare(b.names))
-    setFilteredUsers(filtered)
   }
 
-  const handleAddFriend = async (user: User) => {
-    if (isLoadingFriend[user.id] || user.isFriend || isAnyFriendLoading) return
-    try {
-      setIsAnyFriendLoading(true)
-      setIsLoadingFriend((prev) => ({ ...prev, [user.id]: true }))
-      await api.post("/friends", { friendId: user.id })
-      setAllUsers((prevUsers) =>
-        prevUsers.map((currentUser) =>
-          currentUser.id === user.id
-            ? { ...currentUser, isFriend: true }
-            : currentUser,
-        ),
-      )
-      setFilteredUsers((prevUsers) =>
-        prevUsers.map((currentUser) =>
-          currentUser.id === user.id
-            ? { ...currentUser, isFriend: true }
-            : currentUser,
-        ),
-      )
-      await fetchAllFriends()
-    } catch (error: any) {
-      console.warn("Error adding friend:", error.response.data.message)
-    } finally {
-      setIsAnyFriendLoading(false)
-      setIsLoadingFriend((prev) => ({ ...prev, [user.id]: false }))
-    }
+  const handleAddFriend = (user: User) => {
```

Comment:
let's make this a useCallback

Reply:
We do not need to use useCallback for the handleAddFriend function in this specific context
- handleAddFriend is defined and used entirely within the FindFriends component. It is not being passed down as a prop to child components
- The handleAddFriend function is simple and doesn't involve expensive calculations or memory-heavy operations
- Any changes made by addFriendMutation are handled by React Query's internal cache, making additional memoization redundant. The isPending state is tied to the mutation, ensuring consistent behavior without requiring useCallback
- Using useCallback unnecessarily can lead to bloated and harder-to-read code without measurable performance gains